### ISSUE TITLE
Missing values and MAPL

### ISSUE THREAD

**Issue by mpagowski** [GEOS-ESM/GOCART#252]
NOAA's GBBEPx wildfire files (at 0.1deg x 0.1deg resolution) have missing values. How MAPL treats such values in interpolations?

**Comment by mathomp4** [comment]
@mpagowski Let me ping @bena-nasa and @atrayano to answer this. I'm sure we handle it "correctly" but I'm not sure of the specifics.

**Comment by bena-nasa** [comment]
Are these files that are read via ExtData? Please let me know what version of MAPL you are using.
MAPL has an internal "MAPL undefined" constant MAPL_UNDEF, that we use at various places in the code to protect against operations. It is set to a value of 1.0e15

I don't know what version of MAPL you are using but in newer versions of the generation 2 ExtData, we check the missing value defined in the file, if it does not match the "MAPL undefined" value, read the array from the NetCDF file we set the points that have the file missing value to the "MAPL undefined" one. Then any point in our code base that respects this will be aware of it. For example when regridding the lat-lon grid of the original file to the application grid we respect this.
If the target point has inputs from any points that are the "MAPL undefined" value we do not include them in the application to compute the target value. If all were "MAPL undefined" then the target cell is "MAPL undefined".

Of course if the code or component (outside of MAPL) that ultimately uses these does not protect against doing operations at points that are "MAPL undefined", we cannot control that, the user can do what they want with arrays in their code.

In general our emissions files do not have missing values as how do you handle that in all the components that may use this. Rather if there are no emissions, they are simply 0 since gocart is a huge code base and does not check for any sort of missing value when doing array level operations.


So the answer is that depending on the version of ExtData you are using, we may respect the the file defined missing value and set points that are "missing" to our own internal missing value. This is respected at points in the MAPL code base, but when you get to the application code, all bets are off since we have no control over how the code developer chose to use arrays.

So the ultimate answer is if your input files have missing values, even if MAPL respects them, GOCART does. 

At that point we have 2 options.
1. Reprocess those files so rather than having missing values, they are just set to say 0 at those points.
2. If the above is not an option, but you are using a version of ExtData that respects the file missing value then after the fact in the application code, the arrays would have to be intercepted so that any MAPL_UNDEFs can be set to 0 or protected against
3. If you are using an older version of MAPL where ExtData did not respect the file supplied missing value it was simply not making any accommodation for that we would need to come up with a custom solution

To me of all 3, options seems by far the easiest, it would be a very trivial python script to read in and re-write the file replacing anything that has a missing value with 0.

**Comment by mpagowski** [comment]
Thanks, that answers my question. It is about NOAA's ExtData for wildfires
which come at 0.1deg x 0.1deg resolution
and contain very numerous missing values in places where both wildfires
should not exist or they exist but retrievals are obscured.I can see that
the safest way to deal with this is to convert all those to 0s though that
may not be strictly correct
as it will decrease emissions where the wildfires are burning. Any comments
from NOAA participants/others would be welcome.

**Comment by bena-nasa** [comment]
It doesn't matter what is most "correct", what matters is where is this data used, is it going to ingested and used by an application (I assume GOCART), that does operations of floating point arrays filled from this data. At that point you simply cannot have "missing values" unless EVERY array level operations somehow knows to protect/not use points that are missing, that's just not realistic.

**Comment by mpagowski** [comment]
Yes, it is GOCART, and currently failing so that the data needs to be
preprocessed (i.e. set to 0s) until a solution to distinguish zero
emissions (like ocean) from obscured retrievals is found.

**Comment by bena-nasa** [comment]
Yes, sorry, but by far and away the simplest solution would be to take the existing files and make new versions that have the "undef" points replaced with 0 and just use those. Should be easy enough to do that.

**Comment by mpagowski** [comment]
Yes, that is true but we don't control production of the files

**Comment by bbakernoaa** [comment]
@mpagowski That is partially true.  Yes NESDIS controls the creation of the native files but we can have a preprocessor in the workflow to "fix" the files.

**Comment by mpagowski** [comment]
Yes, we are currently fixing the files for our NRT runs. But the problem
with converting all "missing" to 0s remains

**Comment by bena-nasa** [comment]
Ok, sounds like you can fix this in your workflow. In that you can take the file(s) as produced by NESDIS, make new file(s)  from the originals that have missing value replace by 0, then those are the files that are fed to GOCART. What is the "problem with converting all "missing" to 0s"? Are you asking how one can do this?
